lib/collection/src/collection/point_ops.rs

```diff
@@ -452,17 +453,23 @@ impl Collection {
 
                     CollectionResult::Ok(records)
                 }
-            });
+            })
+            .collect::<FuturesUnordered<_>>();
 
-            future::try_join_all(retrieve_futures).await?
-        };
+        // pre-allocate hashmap with capped capacity to protect from malevolent input
+        let mut covered_point_ids = HashMap::with_capacity(ids_len.min(1024));
+        while let Some(response) = all_shard_collection_requests.try_next().await? {
+            for point in response {
+                // Add each point only once, deduplicate point IDs
+                covered_point_ids.insert(point.id, point);
+            }
+        }
 
-        let mut covered_point_ids = HashSet::new();
-        let points = all_shard_collection_results
-            .into_iter()
-            .flatten()
-            // Add each point only once, deduplicate point IDs
-            .filter(|point| covered_point_ids.insert(point.id))
+        // Collect points in the same order as they were requested
+        let points = request
+            .ids
+            .iter()
+            .filter_map(|id| covered_point_ids.remove(id))
             .collect();
```

Comment:
> Regarding get vs remove, do you propose to get and then clone?

Ah, I didn't consider that it requires a clone then. 

> The de-duplication happens at the time of insertion in the map, only the last write will count

It is, but if we'd be using `get` here that does not work because it still is duplicated in `request.ids`. Not cloning using `remove` makes a lot of sense in which case this isn't a concern anymore.

Reply:
TBH I did not even think about covering the case of duplicated input ids :grin: 
Nice catch, lucky me!